posthog/special_migrations/utils.py

```diff
@@ -0,0 +1,140 @@
+from datetime import datetime
+from typing import Optional
+
+from django.db import transaction
+
+from posthog.celery import app
+from posthog.constants import AnalyticsDBMS
+from posthog.models.special_migration import MigrationStatus, SpecialMigration
+from posthog.special_migrations.setup import DEPENDENCY_TO_SPECIAL_MIGRATION
+
+
+def execute_op(database: AnalyticsDBMS, sql: str, timeout_seconds: int, query_id: str):
+    if database == AnalyticsDBMS.CLICKHOUSE:
+        execute_op_clickhouse(sql, query_id, timeout_seconds)
+        return
+
+    execute_op_postgres(sql, query_id)
+
+
+def execute_op_clickhouse(sql: str, query_id: str, timeout_seconds: int):
+    from ee.clickhouse.client import sync_execute
+
+    sync_execute(f"/* {query_id} */ " + sql, settings={"max_execution_time": timeout_seconds})
+
+
+def execute_op_postgres(sql: str, query_id: str):
+    from django.db import connection
+
+    with connection.cursor() as cursor:
+        cursor.execute(f"/* {query_id} */ " + sql)
+
+
+def process_error(migration_instance: SpecialMigration, error: Optional[str]):
+    update_special_migration(
+        migration_instance=migration_instance,
+        status=MigrationStatus.Errored,
+        last_error=error or "",
+        finished_at=datetime.now(),
+        lock_row=True,
+    )
+
+    from posthog.special_migrations.runner import attempt_migration_rollback
+
+    attempt_migration_rollback(migration_instance)
+
+
+def trigger_migration(migration_instance: SpecialMigration, fresh_start=True):
+    from posthog.tasks.special_migrations import run_special_migration
+
+    task = run_special_migration.delay(migration_instance.name, fresh_start)
+
+    update_special_migration(
+        migration_instance=migration_instance, celery_task_id=str(task.id),
+    )
+
+
+# DANGEROUS! Can cause another task to be lost
```

Comment:
Can you make this a python docstring?
https://www.python.org/dev/peps/pep-0257/